### ISSUE TITLE
Faction GUI can show just one Superweapon type in Command Shortcut

### ISSUE THREAD

**Issue by xezon** [TheSuperHackers/GeneralsGameCode#214]
Faction GUI can show just one Superweapon type in Command Shortcut. This is inconvenient if players owns Superweapons of different types. Boss General can natively build all 3 types, but only the Nuke Missile shortcut is shown.

Ideally the shortcut does not discriminate on Superweapon type, and simply pushes all types either on a single button or on 3 buttons.

**Comment by willherzog** [comment]
It seems to me that the main limitation here is that, currently, each faction can only have up to 11 special power shortcuts defined total, _regardless of how many are actually in use_. It would be great if it could be changed so that, instead, any arbitrary amount of special power shortcuts can be defined (including, for example, those from buildings captured from another faction) and it will simply display as many of those that are in use as will fit on the screen (be that 11 or some new, higher limit to that makes more use of modern resolutions).

**Comment by xezon** [comment]
Indeed, we would need to raise that limit because Airforce General can in theory unlock 11 Powers, and with 2 more it would be 13, over the limit of 11.

**Comment by Foxigod** [comment]
Would bumping up `MAX_SPECIAL_POWER_SHORTCUTS` to some arbitrarily high number, say 64, which we would believe is enough to accommodate both the current special power shortcuts (~14) as well as potential additional shortcuts introduced in custom maps or mods be acceptable? Or do we prefer to swap out the array of fixed size for a vector of dynamic size?

We would in either case then just ensure that we only load as many GUI shortcut buttons as exist in the ControlBar INI file. To take full advantage of this, users would need updated CommandSets that include all the respective shortcuts in their preferred order for each general, and updated ControlBar with more shortcut GUI buttons defined.

**Comment by MTKing4** [comment]
for sanity check, are you guys suggesting it to have multiple columns of support powers buttons? or unlimited columns of them? because that is in no way sustainable, we must have a fixed number size that is enough to accommodate all combinations with a solution in mind, few I can think of are

1. add a scroll buttons to the top and bottom of the screen
2. make super weapons stack over one button, and clicking on it will expand sideways to show all options, (and anything else extra will also be stackable with the other buttons to what fits best for modders who want to add stuff, say they can stack a new power with the A10 button and so on)
3. Make the second column expandable sideways, by adding an arrow button to the side that expands it
4. do the contra mod solution, when having multiple SW, add a button that opens all options in the CommandButton

**Comment by Foxigod** [comment]
I (personally) wasn’t really thinking about anything beyond a “static” UI and thus the need for users to define/download a control bar with more buttons if one wants to take advantage of this, at least in the short term. Thats is, whether you want two columns, smaller icons, wrap around the screen, or whatever else static, that is client-side control bar that needs to be defined.

But I agree that some sort of an accepted solution for UI management when shortcuts become plenty could be beneficial. I.e. some agreement on what sort of a dynamic feature would fit best here.

EDIT: You would, the way I see it, also need an updated INI file[s] for the player templates that define each general’s shortcuts (at least until we possibly eventually drop those shortcut commandset requirements in favour of just a “master” shortcut commandset that is simply generated from all available shortcuts, it that’s preferred).
